sync/class.jetpack-sync-module-users.php

```diff
@@ -54,6 +62,42 @@ public function init_listeners( $callable ) {
 		add_action( 'wp_masterbar_logout', $callable, 10, 0 );
 	}
 
+	public function delay_if_within_wp_insert_user() {
+		if ( ! Jetpack::is_function_in_backtrace( 'do_delayed_calls' ) ) {
+			if ( Jetpack::is_function_in_backtrace( 'wp_insert_user' ) ) {
+				self::$calls_within_wp_insert_user[ current_action() ][] = func_get_args();
+
+				return;
+			}
+		}
+		call_user_func_array( self::$callable, func_get_args() );
+	}
+
+	public function do_delayed_calls() {
+		// We want the save user action to be synced first
+		if ( isset( self::$calls_within_wp_insert_user['jetpack_sync_save_user'] ) ) {
+			$jetpack_sync_save_user = self::$calls_within_wp_insert_user['jetpack_sync_save_user'];
+			unset( self::$calls_within_wp_insert_user['jetpack_sync_save_user'] );
+			self::$calls_within_wp_insert_user = array_merge( array( 'jetpack_sync_save_user' => $jetpack_sync_save_user), self::$calls_within_wp_insert_user );
+		}
+		foreach ( self::$calls_within_wp_insert_user as $action => $calls ) {
+			foreach ( $calls as $arguments ) {
+				if ( 'user_register' === current_filter() ) {
```

Comment:
Consider adding a should_skip_action() function so this code can become more succinct/easier to grok:

``` 
foreach ( self::$calls_within_wp_insert_user as $action => $calls ) {
     foreach ( $calls as $arguments ) {
          if ( self::should_skip_action( $action ) ) {
               break;
          }
          call_user_func_array( 'do_action', array_merge( array( $action ), $arguments ) );
     }
}
```

``` 
function should_skip_action( $action ) {
     if ( 'user_register' != current_filter() ) {
          return false;
     }
     if ( in_array( $action,
          array(
               'jetpack_sync_change_role_user',
               'jetpack_sync_user_locale',
          ), true ) 
     ) {
          return true;
     }
     return false;
}